### PR TITLE
expose: Avoid selector resolution if a selector is not needed

### PR DESCRIPTION
**Pull request by 0xmichalis**
@smarterclayton @kubernetes/kubectl

### CODE REVIEW

File: pkg/kubectl/cmd/expose.go [hunk 1]
```diff
@@ -129,14 +129,18 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
...
-		if kubectl.IsZero(s) {
-			s, err := f.PodSelectorForObject(inputObject)
-			if err != nil {
-				return cmdutil.UsageError(cmd, fmt.Sprintf("couldn't find selectors via --selector flag or introspection: %s", err))
-			}
-			params["selector"] = s
+
+	// For objects that need a pod selector, derive it from the exposed object in case a user
+	// didn't explicitly specify one via --selector
+	if s, found := params["selector"]; found && len(s) == 0 {
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/expose.go]: We set empty params["selector"] in the generator?

Comment (kargakis) [inline pkg/kubectl/cmd/expose.go]: `ParamNames()` returns the desired parameters from the generator and `kubectl.MakeParams(cmd, names)` looks values for those from the flagset. If an object requires a selector and the user doesn't use --selector then `MakeParams` will setup an empty value for "selector".

**Comment by smarterclayton** [comment]
Test case

**Comment by bgrant0607** [comment]
Reassigned to @brendandburns

**Comment by brendandburns** [comment]
Hrm, I don't think this is right.  `params["selector"]` doesn't indicate whether it's needed or not, it indicates if the user specified it or not.

**Comment by kargakis** [comment]
> Hrm, I don't think this is right. params["selector"] doesn't indicate whether it's needed or not, it indicates if the user specified it or not.

MakeParams will scan only flags that are needed: https://github.com/kubernetes/kubernetes/blob/f867ba3ba13e3dad422efd21c74f52b9762de37e/pkg/kubectl/generate.go#L69

Is there anything I am missing here?

**Comment by kargakis** [comment]
> Objects that don't need a selector won't have it in their map.

Added a test case with a test service generator that doesn't need a selector while the user will pass --selector to illustrate this.

**Comment by kargakis** [comment]
I have also slightly refactored unit tests to make them actually work.

**Comment by kargakis** [comment]
I cannot access Shippable. What's the error there?

Tests pass locally for me
```
$ make check WHAT=pkg/kubectl/cmd/
hack/test-go.sh pkg/kubectl/cmd/ 
Running tests for APIVersion: v1,experimental/v1 with etcdPrefix: registry
+++ [0910 13:30:14] Running tests without code coverage
ok  	k8s.io/kubernetes/pkg/kubectl/cmd	0.066s
Running tests for APIVersion: v1,experimental/v1 with etcdPrefix: kubernetes.io/registry
+++ [0910 13:30:22] Running tests without code coverage
ok  	k8s.io/kubernetes/pkg/kubectl/cmd	0.061s
```

**Comment by kargakis** [comment]
I cannot access Shippable. What's the fail about?

**Comment by brendandburns** [comment]
ok, seems reasonable.   Shippable seems like a flake, I kicked it...

**Comment by kargakis** [comment]
Turned out the Shippable error was an artifact in cmd_test.go I carried while cp'ing the service generator.

Everything looks green now

**Comment by smarterclayton** [comment]
LGTM

**Pull request merged**